torchgen/model.py

```diff
@@ -788,9 +796,42 @@ def __post_init__(self) -> None:
                     f"that don't have matching signatures: {test_sig} != {f.func.signature()}"
                 )
         assert self.functional.func.kind() == SchemaKind.functional
+        assert (
+            not self.functional.is_view_op
+        ), "View operator shouldn't be grouped into NativeFunctionsGroup objects. Did you add an out= variant to a view operator?"
         assert self.out.func.kind() == SchemaKind.out
+
         if self.inplace is not None:
             assert self.inplace.func.kind() == SchemaKind.inplace
+            mutable_inputs = [
+                a
+                for a in f.func.arguments.flat_all
+                if a.annotation is not None and a.annotation.is_write
+            ]
+            immutable_rets = [r for r in f.func.returns if r.annotation is None]
+            assert len(mutable_inputs) + len(immutable_rets) == len(
+                self.functional.func.returns
+            ), f"""\
+Found an inplace operator ({self.inplace.func}) with {len(mutable_inputs)} mutable inputs
+ and {len(immutable_rets)} non-aliased return arguments
+ but found a corresponding out-of-place operator ({self.functional.func}) that had an incorrect number of return arguments.
+ Expected {len(mutable_inputs) + len(immutable_rets)} returns, but found {len(self.functional.func.returns)}."""
```

Comment:
The assert here is a little confusing, for two reasons. First, the assert isn't very strong: all it's saying is that the the lengths have to be the same. But the actual inputs and outputs could actually just not have any relationship with each other, but accidentally have the same length. So it seems like, if we're trying to write down what the invariants of this data structure, there is a much stronger statement of this assert.

But should we write this more elaborate assert? This leads to a second confusion: isn't the real correspondence that the `signature()` of all of the variations supposed to be the same. If the signature "preserves" all of the relevant arguments/returns (or, if it drops redundant data, it, itself, asserts the data makes sense), then this is your "stronger" assert. I.e., whatever assert we write here, is a consequence of however signature is defined. So the assert here should be in terms of the signature, somehow.

Reply:
(Done this way, I suspect that this code will be a lot shorter!)

Reply:
I definitely agree with you - the spirit of this check is to make sure that the `signature()`'s of related ops that someone explicitly adds to the yaml match up, and this check makes a pretty mild attempt to capture that (which is complicated and also pretty weak).

It would be awesome if we could just have some assertions that `.signature()` calls are the same, but we don't really know automatically if two related ops have different `.signature()` outputs by accident, because they're intended to - i.e. if there are multiple overloads of an op that should have their own dedicated functional + out= variant.

It would be great if we can figure out somehow by staring at the schema if two different ops were **intended** to be grouped together.

I think that today it's the case that if you ignore `out=` and `TensorOptions` arguments, then if the remaining args + kwargs are the same across two operators, then those ops are expected to be grouped together (and they're `.signature()` should be the same. so if you futz'd up the number of returns in the functional case or something, we'd catch it).


That feels maybe a bit circular, and the check would also have to live somewhere in the grouping logic. But it sounds a lot simpler, and gives you a more understandable error - wdyt?

Reply:
Well, in your brave new world, we would start *deleting* out overloads from `native_functions.yaml` and just rely on them being generated. Then you would know what should be matched, and you'd be generating it yourself anyway so it'd be correct by construction.

Let me put it a little differently. Suppose that someone defines a functional + out variant for op, but they're ever so slightly different, so they don't group up. What kind of error message are you giving today? What can you do in principle? It seems to me that the easiest way to get this going is to have a little field in the yaml that says "I expect to autogen an out for this / I expect it to be grouped" (and you can pick the default one way or another)

Reply:
> have a little field in the yaml that says "I expect to autogen an out for this / I expect it to be grouped" (and you can pick the default one way or another)

this sounds very reasonable. Tentatively, I'm thinking of something like this:
```
- func: _fused_moving_avg_obs_fq_helper(...
  dispatch:
    CPU: fused_moving_avg_obs_fake_quant_cpu
    CUDA: fused_moving_avg_obs_fake_quant_cuda
  # the autogen tag doesn't actually tell the codegen what to generate; it's just used for error checking.
  # codegen will figure out to automatically generate a functional + out= variant based on this op's schema,
  # and then later assert that this `autogen` label was present on the original op
  autogen: functional, out
```